Add a binder... and entirely too much else...

bollwyvl opened the pull request:
Hi! This is a cool looking project!

First off: sorry for this ridiculous PR. In the original scope, I just wanted to push a nice, clean little binder demo PR with an arguably less-hacky way to get the wasm to work (originally from [here](https://github.com/deathbeds/jupyterlab-simple-syntax/blob/master/jupyter_static_syntax/serverextension.py#L25)... we should probably get it into `notebook`, `jupyter_server`, or `jupyterlab_server` so nobody else has to deal). However, going pip-first is going to be great with JupyterLab 3, so it would probably be good to grab that package name,...

But in thinking about the even _larger_ collection of plugins [myst/jupyter-book](https://github.com/executablebooks/meta/discussions/171) would need, it seemed interesting to see what a multi-plugin setup would look like, where the "core" just provided the renderer, and each plugin did its own thing, and could even be extended by other extensions. But this lead to finding things that are supported in the marked renderer today, but missing in vanilla markdown-it, which would be table stakes for taking a real shot at getting something like this into [jupyterlab](https://github.com/jupyterlab/jupyterlab/issues/272) [core](https://github.com/jupyterlab/jupyterlab/pull/499)...

A couple hours later, sure, the [binder](https://gke.mybinder.org/v2/gh/bollwyvl/jupyterlab-markup/add-binder?urlpath=lab%2Ftree%2Findex.ipynb) works, but the remaining code is barely recognizable as i cargo culted things from other extensions.

Not sure how you'd like to proceed, but certainly give the binder a spin!

cc @choldgraf @tonyfast @goanpeca @jasongrout

bollwyvl committed:
start binder

bollwyvl committed:
add serverextension for wasm serving

bollwyvl committed:
ignoring more files, specify python version

bollwyvl committed:
rework with manager, expose plugin machinery, lazy load more

bollwyvl committed:
contributing docs

bollwyvl committed:
rework API, add bootstrap

bollwyvl committed:
add link replacer, options to API, typings in folder

bollwyvl committed:
more docs

bollwyvl committed:
add anchor links, clean up bootstrap build

bollwyvl committed:
try more building

bollwyvl committed:
add legacy header link class

bollwyvl committed:
handle plugins with multiple options

bollwyvl committed:
task lists

agoose77 commented:
Hi @bollwyvl, 

Thank you very much for creating this PR. :superhero:  I created this extension for my own uses during a time when I really oughtn't to have been working on side projects at all :rofl:. Overall, I like the feel of these changes. In particular, eliminating the host mimetypes dependency is a clever trick - when I originally wrote the extension I was unfamiliar with server extensions, and the thought had not occurred to me since that time to monkey-patch the guesser :sparkles:. 

RE the plugin API, I was mulling this over myself because, on the one hand it seems like a lot of development burden for people who just want to load an existing md-it plugin, but on the other, if we just leverage `requirejs` then those extensions are not bundled statically (and increase the loading overhead?). I suppose that it is no bad thing if we have the main plugins bundled in core.

In terms of the design, I'm not an expert in JS - is it good practice to monkey-patch the md renderer instead of creating a new one?

bollwyvl commented:
Thanks for looking!

> RE the plugin API, I was mulling this over myself because, on the one hand
> it seems like a lot of development burden for people who just want to load
> an existing md-it plugin, but on the other, if we just leverage requirejs
> then those extensions are not bundled statically (and increase the loading
> overhead?).
>
Requirejs, and especially the practice of hot loading stuff from CDN, makes
it hard to reason about the final state of the application, and assure the
software will work in the future. The frustrations with this are what
actually lead to building the dependency injection framework we see in lab
today.

In lab 2, there's isn't a good way to achieve lazy loading, short of using
the webpackChunkName mechanism. At least it is a normal ES module load,
rather than a webpack thing.

In lab 3, the situation is on the whole better with the federated
extensions, which makes everything lazy load... But one would still want to
delay that wasm load, preferably until it's actually used, but didn't see
how to do that from the outside. "Pip install" and off you go, and sure all
the backend and frontend are consistent is what we've always wanted.

On the whole, if extensible commonmark becomes a jupyter spec, we're going
to need a way for every document to capture which extensions it needs, and
tools like nbconvert are going to need a way to achieve these things,
preferably without a headless browser... Markdown-it-py is a good step,
svgbob could be _less_ complex... Mermaid would be... harder.

> In terms of the design, I'm not an expert in JS - is it good practice to
> monkey-patch the md renderer instead of creating a new one?
>
Indeed, the change is janky. Not sure what timing gremlins I had found, but
markdown-it wasn't being used at all without this. But on this pr, I was
more interested in finding out what marked features would need to be back
filled with markdown-it plugins. It's likely some upstream API changes
would be needed to do this more cleanly. For a time, it's likely both
engines would need to ship, and be able to switch between them without
editing config files or doing command line stuff.

Feel free to close this PR, if disinterested: it can remain a historical
oddity/reference point if someone wants to continue.

>

agoose77 commented:
@bollwyvl I am really looking forward to the Lab3 mechanism as it will really improve some of the UX issues I've had as both a developer and user of extensions :)

I think I've come across the issues you recall whereby the MD renderer wasn't being used all of the time. 

Overall I'm very happy with this PR, I'll merge it for now because I think it's a significant improvement over where we are currently.

bollwyvl commented:
Feel free to point me to anything you don't like in particular on this PR, happy to help move it forward!  Off the top of my head:

probably in this pr:
- a link to the binder in README
- bump version
  - a full major bump seems reasonable :P

probably in a follow-on (this doesn't need to get any more complicated, unless you want it to):
- handle some additional edge cases that i didn't cover
  - image backgrounds can have preferences
  - `script` tags (gaaaahh)
- validate behavior with some "important" extensions that muck about with markdown, e.g. `jupyterlab-toc`
- a preferences schema that allowed for:
  - turning off unwanted plugins, e.g. `"disabled-plugins": ["whatever"]`
  - or disabling the monkeypatch altogether to see what a document looks like with `marked`
    - if doing automated testing (with e.g. robotframework), this could assess the _actual_ DOM differences without rebuilding
  - probably a minimal ui for config
    - extend `IPluginProvider` with a `title` and `description`, and maybe an `examples` for syntax help

out of scope:
- releasing the serverextension on pypi :P
  - there's also a way to add some breadcrumbs for the extension manager in `package.json/jupyter/lab` (which _could_ be on this PR)
    - but maybe who cares, as lab3 is where it's going? maybe not even release this for lab2?

agoose77 commented:
Yes, agreed on ditching JLab2; once JLab 3 is released I'll upgrade all of the extensions I use on a daily basis where I can :)

Your list of issues looks like a great place to start. I'd suggest the separation into a new PR as you mention :)

One additional feature that *might* be a goal for the follow-up PR is allowing this parent extension to configure the plugins by passing in an options object. I don't know if it is considered reasonable to pass non-schema-backed configurations from JLab configuration to the consumer, but it would mean that users could easily configure things.

bollwyvl committed:
per-plugin options

bollwyvl committed:
refactor builtins

bollwyvl committed:
slim down CONTRIBUTING

bollwyvl committed:
add settings boilerplate

bollwyvl committed:
basic style, table

bollwyvl committed:
add plugin disabling

bollwyvl committed:
use article/section instead of table

bollwyvl committed:
some more style tweaks

bollwyvl committed:
fix nav

bollwyvl committed:
fix rel

bollwyvl committed:
provide advanced settings escape hatch

bollwyvl committed:
add default value to markdownit options

bollwyvl commented:
>  parent extension to configure the plugins by passing in an options object

Hm, didn't see your response for some reason, but independently built that out with a settings/doc form. I can force push back to 7ca1f71, and start a new branch....

agoose77 commented:
> > parent extension to configure the plugins by passing in an options object
> 
> Hm, didn't see your response for some reason, but independently built that out with a settings/doc form. I can force push back to [7ca1f71](https://github.com/agoose77/jupyterlab-markup/commit/7ca1f71161f8ca9ff8680e546d4edaf581f02df2), and start a new branch....

Don't worry - having thought about it some more, being able to configure the plugins is reasonably important such that I think it should be included in the same PR.

Thanks for putting in all these hours, I'm still playing catch up on my week!

bollwyvl commented:
> should be included in the same PR.

Cool. Short of some more feedback, this is probably in a mergeable state (provided it looks good on binder)... maybe I can rope some more folk in for playtesting. Feel free to squash merge, as there's a lot of noise, as i tend towards very small, incremental commits. I'll make some issues for the follow-on work.

> Thanks for putting in all these hours

It's been a pleasure. Even though a lot of these things are annoying, I've been down this road a number of times, and am happy to save some scars for folks. The ideas in the repo (to begin with) are important, and needs to be shown for real vs. more talking about them in the abstract, which is all we've seen for a long time!

> playing catch up on my week!

Tell me about it... time just gets away from me on a lot of these things!

choldgraf commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,34 @@
+## Extending
+
+Additional [`markdown-it` plugins][plugins]
+can be added as small labextensions. After getting started with the [official cookiecutter][],
```
this will be much easier once we can `pip install` labextensions!

bollwyvl replied:
for _users_, yes. _contributors_ will need _even more_ webpack stuff, and will have to track to a specific version of a jupyterlab-specific package. however, the dev docs for this repo will still be pretty simple, with the addition of `jupyter labextension build .`.

choldgraf commented:
a few thoughts:

1. Wow, this is awesome
2. That little "demo" page to show which extensions are active is really cool!
3. The main thing I think this needs is more documentation. I had questions like:
  - How do I know to look for that demo page?
  - We should document how others can add new markdown-it extensions
  - A list of the extensions that are available in the docs (outside of the JupyterLab demo page) would be helpful too

Also trying to open a "regular" markdown file with markdown preview resulted in this error:

![image](https://user-images.githubusercontent.com/1839645/98047512-a192ac00-1df1-11eb-8952-cf86254c9d9b.png)


In general, this is really cool, you're both wizards

agoose77 commented:
> a few thoughts:
> 
> 1. Wow, this is awesome
> 2. That little "demo" page to show which extensions are active is really cool!
> 3. The main thing I think this needs is more documentation. I had questions like:
> 
> * How do I know to look for that demo page?
> * We should document how others can add new markdown-it extensions
> * A list of the extensions that are available in the docs (outside of the JupyterLab demo page) would be helpful too
> 
> Also trying to open a "regular" markdown file with markdown preview resulted in this error:
> 
> ![image](https://user-images.githubusercontent.com/1839645/98047512-a192ac00-1df1-11eb-8952-cf86254c9d9b.png)
> 
> In general, this is really cool, you're both wizards

Thanks for the feedback - I'll take a look tomorrow (British time!) 🙂

bollwyvl commented:
Thanks for the review!

> How do I know to look for that demo page?

the binder link? but again (#13), once un-spec'd syntax extensions enter the picture, it's not going to look like much... and indeed look broken, as github can barely render _regular_ notebooks, and nbviewer will just :shrug:. 

> The main thing I think this needs is more documentation.

primary focus of this PR is getting to something that even _sorta_ works, and reproduces the existing behavior.

> A list of the extensions that are available in the docs

From a docs perspective, it is "above the fold" in the README, and then would be expanded with screenshots? but there's also the `Markdown Settings` entry (right now, just through the command palette). As mentioned on #13, there's actually a bunch of (live-updating) aspects that an author/reader needs to know:
- what extensions are installed
- what extensions are enabled
- what extensions are currently in use _by this document_

I'm kinda thinking this ends up in the document toolbar... plain text markdown has one, it's just usually collapsed.

> We should document how others can add new markdown-it extensions

Probably a branding issue, but making new plugins for markdown-it _itself_ is _entirely_ outside the scope... but if we're talking _integrating an existing  markdown-it plugin to lab_, at present, the best i can offer is, as stated there:
- run cookiecutter
- change `plugin.ts` to look like one of the examples

#11 will be a good place to land a proper tutorial, while a (continuously merged, tested) cookiecutter fork would be my preference for the "truth" of how to write a greenfield extension.

> trying to open a "regular" markdown file 

good catch! taking a look, presumably i did something bad with my ugly hack to get a relative URL synchronously. Basically _nothing_ in markdown-it can ever [return a promise](https://github.com/markdown-it/markdown-it/blob/master/docs/development.md#i-need-async-rule-how-to-do-it), and the url resolver, which has to account for things like alternate contents managers, etc. is async. It's probably worth thinking about how plugins might get access to rendering lifecycle events, e.g. `postrender`, `postlatex`, and probably a lot more.

> you're both wizards

trying to take some of the magic _out_ of it!

choldgraf commented:
Thanks for these responses @bollwyvl - just to be clear, I don't think any of the docs stuff should be a _blocker_ on that, as long as @agoose77 likes the direction of this PR and thinks it keeps things moving in the right direction, I am all for it. Mostly I am just trying to be helpful where I can, which is to serve as a "person that doesn't understand JupyterLab that well and will probably be semi-representative of a first-time user", thus my focus on docs...but that can certainly be in a different PR and I'd be happy to help with it once I have my head wrapped around this extension :-)

agoose77 commented:
@bollwyvl yeah the async thing is a bit of a pain. Reading the [md-it wiki](https://github.com/markdown-it/markdown-it/blob/master/docs/development.md#i-need-async-rule-how-to-do-it), their suggestion is a two pass approach which should be simple enough. I'll add it to my todo :)

goanpeca commented:
This is wawesome work! 🥳

bollwyvl committed:
remove redundant link replacer, add main menu entries, some docs

bollwyvl commented:
Turns out, through the magic of reuse, the part of the stock renderer we repurpose but _don't_ overload just does the link handling already, so I've clobbered the `markdown-it-replace-link` thing. This sidesteps the issue... for now. Some lifecycle hooks are probably a good idea.

Just pushed with that, as well as some low-hanging fruit from review:
- a _Usage_ section in README
- adding some main menu items

I'm going to hold off on the _third_ level of discoverability (toolbar button), as there are some design decisions that need to be shopped a bit more on #13...

bollwyvl commented:
One thing i did note (waiting for binder to build): the `markdown-it-anchor` slugify approach is not the same as lab/notebook/nbviewer... otherwise looking better than yesterday!

agoose77 commented:
> One thing i did note (waiting for binder to build): the `markdown-it-anchor` slugify approach is not the same as lab/notebook/nbviewer... otherwise looking better than yesterday!

Actually this reminds me, we should check this with JLab ToC - IIRC there might be some issues related to the anchor slugs. What do you say to merging this here (with the proviso that the anchors work!) and then rebasing the rest upon master?

agoose77 commented:
Also, @bollwyvl is there a reason that you're not included in the LICENSE file?

bollwyvl commented:
> not included in the LICENSE file?

I'm good... happy to do grunt work, if it enables others to do more with less work... and handle on-going maintenance! But seriously, I think the big goal is to mature this to where it could be considered for upstreaming into core. It's a lot easier to make the case with a robust, thought-out, already-works-today implementation than a long-running PR (oops).

bollwyvl committed:
lab-compatible anchor slugs, docs, add toc to binder, linting, clean up server ext

bollwyvl committed:
add binder links

bollwyvl committed:
more badges

bollwyvl commented:
Pushed with a fix to anchor URLs, verified while watching binder load.

Also added:
- quick test of `display` to notebook (works)
- added `@jupyterlab/toc` to binder (looks good locally)
  - though `toc` eats the pilcrow :cry: 
- package/binder badges to readme (mostly broken links for now

I'd say kick the tires once more, then we're ready to rock, think about how to proceed.

bollwyvl commented:
Yeah, binder looks good... toc doesn't add numbers to markdown files with either renderer, so not-bollwyvl's-problem...

choldgraf commented:
This looks great to me 👍

agoose77 commented on src/settings.tsx:
```diff
@@ -0,0 +1,280 @@
+              {provider.title}
+            </label>
+          </h4>
+          <ul className={DOCS_CLASS}>{docs}</ul>
```
So, this might seem like bikeshedding, but how would you feel about making this `ul` inline?
E.g. with 
```css
ul {
flex-direction: row;
flex-wrap: nowrap;
}
``` 

![image](https://user-images.githubusercontent.com/1248413/98219367-ce0b0d00-1f44-11eb-8c92-fd0cd83d295e.png)

bollwyvl replied:
good catch, that was the intent. i guess i changed it to try to handle narrow views, but it still looks pretty good down to about 700 pixels. i guess being able to collapse the navigation would be the only predictable way to save space...

pushed d57f382

agoose77 commented on src/plugin.ts:
```diff
@@ -0,0 +1,122 @@
+          settingsMain.title.icon = markupIcon;
+          settingsMain.disposed.connect(() => (settingsMain = null));
+        }
+        shell.add(settingsMain, 'main');
```
Is `.add` idempotent, or should it be behind the null guard? This is a learning point for me rather than looking for bugs at this stage :)

bollwyvl replied:
`add` has the side-effect of activating the widget: so if it's been opened, but you pop to single document mode, and then hit it (which would probably not work in lab3 due to its crazypants behavior) it will just show it rather than trying to re-create it.

agoose77 commented on src/plugin.ts:
```diff
@@ -0,0 +1,122 @@
+      caption: 'Reopen documents to see changes',
+      isToggled: () => enabled,
+      isEnabled: () => manager.settings != null,
+      execute: async (args) => {
```
Can this `async` keyword be removed?

agoose77 replied:
@bollwyvl do you have any thoughts here?

bollwyvl replied:
sorry, yes, just removed... was left over from when it might have been a promise on settings, but the cache takes care of it.

agoose77 replied:
Ah, my mistake. Thanks :)

bollwyvl replied:
... only because you reminded me!

agoose77 commented:
These are some really cool changes - I've not delved into the `VDom` support yet, so it is nice to see its utility in a larger extension. Overall, I'm really happy with this PR - I think this lays the groundwork for a really positive addition to the JLab ecosystem.

One thing I thought about when writing this extension is the current state of maths rendering - right now we have JLab extensions to change the maths renderer. This extension creates an additional extensibility layer one level lower, in the MD renderer itself via md-it plugins. However, right now we still defer to JLab for maths processing, so md-it cannot influence this stage. Should we move to an entirely md-it implementation and leverage md-it plugins for md rendering? This would require creating an additional MathJax + KaTeX extension.

bollwyvl committed:
unwrap docs links

bollwyvl commented:
> VDom support 

I'm no great fan of react (and actively dislike jsx, as i carry many scars from CFXML), but it's the best shipped thing we have for leaf nodes since the Licensing Fiasco was resolved, and likely not to be going anywhere. the `lumino/vdom` stuff was also fine, but nothing in particular works with it (even the built-in lab stuff). but you can read it in an afternoon, vs react which is like a municpal water supply... few really understand how it works, but you all have to share one: when it works, its fine. when it breaks the whole city shuts down.

such a panel should _probably_ be several lumino widgets (nav, sidebar, etc). but the scope of this just isn't large enough to warrant it, even though that file's getting pretty long. the win there would be, for example, each of the examples could _be_ a notebook cell, which one could drag directly into notebook.

>  MathJax + KaTeX extension.

not poking that particular bear on this PR! that _any_ latex _works_ (like the link replacing) gives me cause for joy. people get _real_ particular about their maths, so i haven't even looked at katex, and i don't know at whose feet it lies to document the inconsistencies... not me!

that being said, as some folks _don't_ use math, and it does cause rendering hiccups, adding another layer of options to allow disabling math typesetting altogether would lie on this extension. this would allow a plugin to add their own and not have them fight, without putting the burden here. frankly, i'd like to see an extension that showed the performance (which would be bad) vs fidelity (which would be perfect) of deferring to a backend renderer, serving (cached), clean-as-the-driven snow SVG... for certain folks, this is the only real answer, as they have all their special macro packages, etc. for their field that nobody is going to take the time to port to js.

agoose77 commented:
I hadn't really thought about the different VDom implementations that are out there, interesting. Probably out of discussion scope for this PR, but I'm sure you'll hear more questions from me on the subject. 

>i'd like to see an extension that showed the performance (which would be bad) vs fidelity (which would be perfect) of deferring to a backend renderer, serving (cached), clean-as-the-driven snow SVG... for certain folks, this is the only real answer, as they have all their special macro packages, etc. for their field that nobody is going to take the time to port to js.

This is actually a fascinating idea. There are several packages that I'd like but can't use at the moment... and a really dumb implementation wouldn't be that hard at all using processes. Food for thought ... :fire:

bollwyvl commented:
Ah, so you're one of them! It's good we're having these discussions, and
have different perspectives coming in to it!

That behavior raises the bar on the lifecycle stuff, and complicates the
server-side stuff, but svg is archival. Over on jupyterlab-lsp, we've been
using texlab/tectonic stack for testing, and it's been... interesting.

So then this extension might _also_ need metadata documenting the tex
packages needed, and the simple list would be insufficient.

choldgraf commented:
So this conversation is now in a language that I no longer understand, but FWIW I support both of you continuing to be wizards and doing whatever you think is best 😅

agoose77 commented:
I don't think a tectonic renderer would need to include any metadata about packages - doesn't it support lazy loading of packages just like TeXLive does? (I only briefly read the docs).

bollwyvl commented:
> doesn't it support lazy loading of packages

oh and how! i trust tex stuff more than, say, unpkg/npm/requirejs, but am not sure why... but it's definitely another source of irreproducibility, or at least bewildering initial delay.

jasongrout commented:
> react which is like a municpal water supply... few really understand how it works, but you all have to share one: when it works, its fine. when it breaks the whole city shuts down.

I had to chuckle with your analogies here.

agoose77 commented:
> > doesn't it support lazy loading of packages
> 
> oh and how! i trust tex stuff more than, say, unpkg/npm/requirejs, but am not sure why... but it's definitely another source of irreproducibility, or at least bewildering initial delay.

If we require users to install a separate extension to handle "true" LaTeX rendering, I think a disclaimer about initial package-load delays should be fine. In my experience using LaTeX distributions, the automatic package management approach has been a far more user friendly experience.

Also, whilst the information declaring which *md-it* plugins are required to render a notebook are not stored in the markup, but instead in the notebook metadata, it should not constrain future modifications / improvements to the specification. Understandably we want to get this right, but I think that would give us more leeway vs needing to anticipate ahead :)

bollwyvl commented:
Yeah, just-in-time resolution-of-gigabytes-from-the-internet is great until.... it isn't. I'm in the `pip/conda/apt install foo` camp: take as long as you want, let me check in a fully-realized solution description from sources I know and trust (if even unfounded), and keep working when i get on a plane (remember when we got on planes?), and be able to reproduced on another machine at a later date. 

ANYHOOOO: actionable takeaway of the above is: 
- the core extension can (but not on this PR, folk think it's worth it):
  - allow disabling built-in math typesetting (and or other features, like link replacement)
  - provide access to lifecycle hooks to plugins for when those features would have happened
- a separate repo/extensoon entirely, can:
  - make use of the above
  - add a dedicated REST API (or perhaps, websocket, if it might be minutes before it "succeeds")
  - configurably source rendering from wherever it needs to in a thread
  - provide same at `nbconvert` time

agoose77 commented:
> Yeah, just-in-time resolution-of-gigabytes-from-the-internet is great until.... it isn't. I'm in the `pip/conda/apt install foo` camp: take as long as you want, let me check in a fully-realized solution description from sources I know and trust (if even unfounded), and keep working when i get on a plane (remember when we got on planes?), and be able to reproduced on another machine at a later date.
> 
> ANYHOOOO: actionable takeaway of the above is:
> 
> * the core extension can (but not on this PR, folk think it's worth it):
>   
>   * allow disabling built-in math typesetting (and or other features, like link replacement)
>   * provide access to lifecycle hooks to plugins for when those features would have happened
> * a separate repo/extensoon entirely, can:
>   
>   * make use of the above
>   * add a dedicated REST API (or perhaps, websocket, if it might be minutes before it "succeeds")
>   * configurably source rendering from wherever it needs to in a thread
>   * provide same at `nbconvert` time

Hehe, I think that's where having a global package cache becomes a "feature" :P

I think your points summarise the issue nicely :)

bollwyvl committed:
remove async from enabled check because cache

choldgraf commented:
Can I make a recommendation that we triage this PR's comments for some issues to create as future discussion / action points, and then get this PR shipped? :shipit:

:-)

agoose77 commented:
I've been having some more thoughts about this @bollwyvl 

Are you thinking that we store the IDs of the JupyterLab Markdown Plugin Extensions somewhere, rather than the IDs of the MarkdownIt plugins themselves? I was initially thinking of some kind of system where the MarkdownIt extension checks whether it has the providers for requested plugins, but this would tie the metadata source (e.g. notebook JSON or markdown header) to the implementation details (that we use MarkdownIt). 

I was thinking that it would be better for the metadata source to request JLab extensions. If this were the case, at what point is it not better to generalise this approach and have notebooks be able to suggest which frontend extensions they expect. This wouldn't be a hard requirement, because people use all kinds of notebook frontends, and there are often different extensions to implement the same functionality. But, at least in JupyterLab this might be quite useful, e.g. a notebook could state

I need the following extensions:
* jupyterlab-diagrams
* ipympl

But, then we arrive at the point for JLab 3 where these extensions are already managed by the Python dependency management, and the whole thing becomes a lot simpler. I know that you can still load dependencies from npm etc., but from the "good notebook workflow" perspective, the standard approach is to use a `requirements.txt` or conda `environment.yml` to capture Python dependencies; there is a precedent for reproducibility. 

TL;DR, is it sufficient to implicitly capture md-it plugin requirements using the Python dependency manager?

agoose77 commented:
> Can I make a recommendation that we triage this PR's comments for some issues to create as future discussion / action points, and then get this PR shipped? :shipit:
> 
> :-)

I think @bollwyvl has already managed to do that :) I've moved my further discussion there now :)

agoose77 commented:
I think this PR as-is is worth merging. The additional points of discussion, such as supporting codemirror extension need to be addressed before making a new release, but I'd prefer to do that from a new PR.

Another big thank-you to all contributors, especially @bollwyvl, for the work in this PR.

choldgraf commented:
🎉🎉🎉🎉 this is very exciting :-)  another step closer to much-improved markdown in JupyterLab!

agoose77 commented:
@choldgraf if you're interested, I've added some alpha versions to PyPI and npm so that you can install it as described in the README.

I'm waiting on JupyterLab 3.0 to support installing only via the PyPI package.
